target_postgres/sinks.py

```diff
@@ -142,35 +152,74 @@ def bulk_insert_records(  # type: ignore[override]
-        data_to_insert: List[Dict[str, Any]] = []
+        data_to_insert: Tuple[Tuple[Any]] = None
 
         if self.append_only is False:
-            insert_records: Dict[str, Dict] = {}  # pk : record
+            copy_values: Dict[str, Tuple] = {}  # pk : values
             for record in records:
-                insert_record = {}
-                for column in columns:
-                    insert_record[column.name] = record.get(column.name)
+                values = tuple((record.get(column.name) for column in columns))
                 # No need to check for a KeyError here because the SDK already
                 # guaruntees that all key properties exist in the record.
                 primary_key_value = "".join([str(record[key]) for key in primary_keys])
-                insert_records[primary_key_value] = insert_record
-            data_to_insert = list(insert_records.values())
+                copy_values[primary_key_value] = values
+            data_to_insert = tuple(copy_values.values())
         else:
-            for record in records:
-                insert_record = {}
-                for column in columns:
-                    insert_record[column.name] = record.get(column.name)
-                data_to_insert.append(insert_record)
-        connection.execute(insert, data_to_insert)
+            data_to_insert = [
+                tuple((record.get(column.name) for column in columns))
+                for record in records
+            ]
+
+        # Prepare to process the rows into csv. Use each column's bind_processor to do
+        # most of the work, then do the final construction of the csv rows ourselves
+        # to control exactly how values are converted and which ones are quoted.
+        column_processors = [
+            column.type.bind_processor(connection.dialect) or str for column in columns
+        ]
+
+        def process_column_value(data: Any, proc: Callable) -> str:
+            # If the data is null, return an unquoted, empty value.
+            # Unquoted is important here, for PostgreSQL to interpret as null.
+            if data is None:
+                return ""
+
+            # Pass the Python value through the bind_processor.
+            value = proc(data)
+
+            # If the value is a string, escape double-quotes as "" and return
+            # a quoted value.
+            if isinstance(value, str):
+                # escape double quotes as "".
+                return '"' + value.replace('"', '""') + '"'
+
+            # If the value is a list (for ARRAY), escape double-quotes as \" and return
+            # a quoted value in literal array format.
+            if isinstance(value, list):
+                # for each member of value, escape double quotes as \".
+                return (
+                    '"{'
+                    + ",".join('""' + v.replace('"', r'\""') + '""' for v in value)
+                    + '}"'
+                )
```

Comment:
I think once the tests are passing, we're good to go here. This method of escaping arrays stands out as awkward, but I think it's necessary. I tried putting together an implementation that avoids it using a custom dialect of csv writer. But I think that anything I come up with there is going to be even more janky.

Thoughts @visch ?

Reply:
That makes sense, surprising but if this is better that makes sense to me!

```
CREATE TABLE public.testing_table (
    id SERIAL PRIMARY KEY,
    jsonb_data JSONB[]
);

INSERT INTO public.testing_table --succeeds, but breaks backward compatibility
(jsonb_data)
VALUES('{"[\"string1\", \"string2\", 1, 2, 3]"}'::jsonb[]);

INSERT INTO public.testing_table --succeeds
(jsonb_data)
VALUES('{\"string1\", \"string2\", 1, 2, 3}'::jsonb[]);

INSERT INTO public.testing_table --fails
(jsonb_data)
VALUES('[\"string1\", \"string2\", 1, 2, 3]'::jsonb[]);

--ERROR: malformed array literal: "[\"string1\", \"string2\", 1, 2, 3]"
--  Detail: "[" must introduce explicitly-specified array dimensions.

INSERT INTO public.testing_table --fails
(jsonb_data)
VALUES('"[\"string1\", \"string2\", 1, 2, 3]"'::jsonb[]);

--ERROR: malformed array literal: ""[\"string1\", \"string2\", 1, 2, 3]""
--  Detail: Array value must start with "{" or dimension information.
```

@kinghuang

Reply:
@kinghuang are you interested in getting this PR to the finish line? Is there anything I can do to help?

> I'll check those out. The [csv.QUOTE_NOTNULL](https://docs.python.org/3/library/csv.html#csv.QUOTE_NOTNULL) and [csv.QUOTE_STRINGS](https://docs.python.org/3/library/csv.html#csv.QUOTE_STRINGS) would both be more convenient, but they are new in Python 3.12.

Would it be worth using those to make the tests pass on Python 3.12? If they significantly simplify the implementation, we can think about backporting the `csv` library from Python 3.12+ and publish the wheels to PyPI.